src/pause/138CC0.c

```diff
@@ -103,19 +103,386 @@ s32 pause_badges_scroll_offset_x(s32 x) {
     return x;
 }
 
-INCLUDE_ASM(s32, "pause/138CC0", pause_badges_try_remove);
+s32 pause_badges_try_remove(s16 badgeID) {
+    s16 *currentSlot = gPlayerData.equippedBadges;
+    s16 *slotToRemove = currentSlot;
+    s32 result = 0;
+    s32 i;
+
+    if (badgeID == 0x7FFF) {
+        return 0;
+    }
+
+    for (i = 0; i < ARRAY_COUNT(gPlayerData.equippedBadges); ++i, ++currentSlot) {
+        if (badgeID == *currentSlot) {
+            slotToRemove = currentSlot;
+            result = 1;
+            break;
+        }
+    }
+
+    // Remove and shift array contents
+    for (; i < ARRAY_COUNT(gPlayerData.equippedBadges)-1; ++i, ++currentSlot) {
+        *currentSlot = *(currentSlot+1);
+    }
+    *currentSlot = 0;
+
+    if (*slotToRemove == 0) {
+        result = 2;
+    }
+    return result;
+}
+
+typedef s32 BadgeEquipResult;
+enum BadgeEquipResults {
+    EQUIP_RESULT_ALREADY_EQUIPPED,
+    EQUIP_RESULT_NOT_ENOUGH_BP,
+    EQUIP_RESULT_TOO_MANY_BADGES,
+    EQUIP_RESULT_SUCCESS,
+};
+
+BadgeEquipResult pause_badges_try_equip(s16 badgeID) {
+    PlayerData* playerData = &gPlayerData;
+    s16 *badgeSlot = &playerData->equippedBadges[0];
+    s32 i;
+    s32 totalEquippedBP;
+
+    if (badgeID == 0x7FFE) {
+        badgeID = 0;
+    }
+    if (badgeID == 0x7FFF) {
+        badgeID = 0;
+    }
+    if (badgeID == 0) {
+        return EQUIP_RESULT_NOT_ENOUGH_BP;
+    }
+
+    for (i = 0; i < ARRAY_COUNT(playerData->equippedBadges); ++i, ++badgeSlot) {
```

Comment:
`i++, badgeSlot++` - also, have you tried not using badgeSlot + co and just using thing[i] syntax for these loops?

Reply:
Yeah I have, the obvious `badgeSlot[i]` messes with register allocation a bit (relocates `badgeSlot` from `a1` to `v1`, probably something to do with pushing the final write to `badgeSlot` farther up in the function), I couldn't find a better way to fix it.

Reply:
preincrement -> postincrement is done

Reply:
kk thanks!